frontend-react/src/hooks/UseValueSets.ts

```diff
@@ -0,0 +1,152 @@
+// value set table name
+// we may not need the generic here, react-query probably handles that. But it couldn't hurt?
+const getLatestData = <T>(tableName: LookupTables, version: number) => {
+    const endpointHeader = lookupTableApi.getTableData<T>(version, tableName);
+
+    return axios
+        .get<T>(endpointHeader.url, endpointHeader)
+        .then((response) => response.data);
+};
+
+// not going to type this yet, as this function will likely go away when we bring in fetch by provider
+// returns a list of all lookup tables to be filtered
+const getLookupTables = () =>
+    axios(lookupTableApi.getTableList()).then(({ data }) => data);
+
+/*
+
+  Helper function to find the table we want within the response for ALL tables
+  Hopefully this will go away with the API refactor
+
+*/
+const findTableByName = (
+    tables: LookupTable[],
+    tableName: LookupTables
+): TableAttributes => {
+    const filteredBody: LookupTable[] = tables.filter(
+        (tv: LookupTable) => tv.tableName === tableName && tv.isActive
+    );
+
+    if (!filteredBody.length) {
+        throw new Error(`Table '${tableName}' was not found!`);
+    }
+    const table: LookupTable = filteredBody.sort(
+        (a: LookupTable, b: LookupTable) =>
+            b["tableVersion"] - a["tableVersion"]
+    )[0];
+
+    const { tableVersion, createdAt, createdBy } = table;
+    if (tableVersion === undefined) {
+        throw new Error(`No version of table '${tableName}' was found!`);
+    }
+
+    return {
+        version: tableVersion,
+        createdAt,
+        createdBy,
+    };
+};
+
+/*
+
+  useQuery function
+
+  hook used to get value sets and value set rows (defined by passsed dataTableName)
+
+*/
+
+export const useValueSetsTable = <T extends ValueSet | ValueSetRow>(
+    dataTableName: LookupTables,
+    suppliedVersion?: number
+): {
+    valueSetArray: Array<T>;
+    error: any;
+} => {
+    let versionData: Partial<TableAttributes> | null;
+    let error;
+    let valueSetArray;
+
+    // get all lookup tables
+    const { error: tableError, data: tableData } = useQuery<LookupTable[]>(
+        ["lookupTables"],
+        () => getLookupTables(),
+        { enabled: !suppliedVersion } // only if version was not already passed in
+    );
+
+    // handle complexity around passed version vs. incomplete useQuery vs. success
+    if (suppliedVersion) {
+        versionData = {
+            version: suppliedVersion,
+        };
+    } else if (!tableData) {
+        versionData = null;
+    } else {
+        try {
+            versionData = findTableByName(tableData, dataTableName);
+        } catch (e) {
+            error = e;
+            versionData = null;
+        }
+    }
+
+    // unclear how to resolve the issue of defining a function to account for the absence of
+    // version when the `enabled` param should take care of not running if that's not present...
+    // get value for lookup table version
```

Comment:
It doesn't seem like we'd need the extra safety 🤷🏼‍♂️ If I understand, we're not calling the first query if a supplied version is given...which then could possibly be `null` resulting in the second query not calling? Would that result in us getting back `{ error: undefined, valueSetArray: undefined }`? If so, the return type should be updated to match that possibility.

Reply:
a few things to mention here:
* my first pass at this function included early returns. This got complex in part because with the hooks rules, we need to call all useQuery hooks in all invocations. The better pattern seemed to be what I have here where we define the return variables (error and data, or in this case `valueSetArray`) up top, and then mess with them during the course of the function, making sure that by the end of the function they are defined the way we want them before returning.
* as such, I don't think there's a possibility of returning `undefined` values here. The function will always reach the end, and at the end we always define the values one way or another
* the complexity I was getting at in this comment is around needing to pass that `Promise.reject` function to handle the case of `versionData?.version` not being defined, even though we're explicitly telling useQuery not to run the function if it's not defined. There may be a better way to do this, but I couldn't find one.

Reply:
Ahhh, okay, I get the flow a little better now. So the _first_ call up top is the conditionally-run call based on if there's version data or not already supplied. If not, it gets it for us. But the call below, where you have `Primse.reject` _always_ runs. Error uses the right error, valueSetArray creates the array and defaults to empty.

I'll add a suggestion to try out but no biggie if that doesn't work